drivers/can/can_xmc4xxx.c

```diff
@@ -0,0 +1,998 @@
+{
+	uint32_t reg = fifo->base->MOFGPR;
+	uint8_t top_index, bot_index, cur_index;
+	uint8_t head_index = FIELD_GET(CAN_MO_MOFGPR_CUR_Msk, reg);
+
+	fifo->head = &CAN_MO->MO[head_index];
+	top_index = FIELD_GET(CAN_MO_MOFGPR_TOP_Msk, reg);
+	bot_index = FIELD_GET(CAN_MO_MOFGPR_BOT_Msk, reg);
+	cur_index = FIELD_GET(CAN_MO_MOFGPR_CUR_Msk, reg);
+
+	LOG_DBG("Fifo: top %d, bot %d, cur %d", top_index, bot_index, cur_index);
+}
+
+static void can_xmc4xxx_rx_fifo_handler(const struct device *dev, struct can_xmc4xxx_rx_fifo *fifo,
+					struct can_xmc4xxx_rx_callback *rx_callback)
+{
+	bool is_rtr = (fifo->base->MOSTAT & CAN_MO_MOSTAT_DIR_Msk) != 0;
+
+	while (!can_xmc4xxx_is_fifo_empty(fifo)) {
+		struct can_frame frame;
+		CAN_MO_TypeDef *mo_tail = fifo->tail;
+
+		memset(&frame, 0, sizeof(frame));
+
+		if ((mo_tail->MOAR & CAN_MO_MOAR_IDE_Msk) != 0) {
+			frame.flags |= CAN_FRAME_IDE;
+			frame.id = FIELD_GET(CAN_MO_MOAR_ID_Msk, mo_tail->MOAR);
+		} else {
+			frame.id = FIELD_GET(XMC_CAN_MO_MOAR_STDID_Msk, mo_tail->MOAR);
+		}
+
+		frame.dlc = FIELD_GET(CAN_MO_MOFCR_DLC_Msk, mo_tail->MOFCR);
+
+		if (!is_rtr) {
+			memcpy(&frame.data[0], (void *)&mo_tail->MODATAL, sizeof(uint32_t));
+			memcpy(&frame.data[4], (void *)&mo_tail->MODATAH, sizeof(uint32_t));
+		} else {
+			frame.flags |= CAN_FRAME_RTR;
+			memset(&frame.data[0], 0, CAN_MAX_DLEN);
+		}
+
+		if (rx_callback->function != NULL) {
+			rx_callback->function(dev, &frame, rx_callback->user_data);
+		}
+
+		/* reset the rx pending bit on the tail */
+		mo_tail->MOCTR = XMC_CAN_MO_RESET_STATUS_RX_PENDING;
+		can_xmc4xxx_increment_fifo_tail(fifo);
+	}
+}
+
+static void can_xmc4xxx_rx_handler(const struct device *dev)
+{
+	struct can_xmc4xxx_data *dev_data = dev->data;
+
+	for (int i = 0; i < CONFIG_CAN_MAX_FILTER; i++) {
+		if ((BIT(i) & dev_data->filter_usage) == 0) {
+			continue;
+		}
+
+		can_xmc4xxx_update_fifo_head(&dev_data->rx_fifos[i]);
+		can_xmc4xxx_rx_fifo_handler(dev, &dev_data->rx_fifos[i],
+					    &dev_data->rx_callbacks[i]);
+#if defined(CONFIG_CAN_ACCEPT_RTR)
+		can_xmc4xxx_update_fifo_head(&dev_data->rtr_fifos[i]);
+		can_xmc4xxx_rx_fifo_handler(dev, &dev_data->rtr_fifos[i],
+					    &dev_data->rx_callbacks[i]);
+#endif
+	}
+}
+
+static void can_xmc4xxx_state_change_handler(const struct device *dev, uint32_t status)
+{
+	const struct can_xmc4xxx_config *dev_cfg = dev->config;
+	struct can_xmc4xxx_data *dev_data = dev->data;
+	enum can_state new_state;
+	struct can_bus_err_cnt err_cnt;
+
+	can_xmc4xxx_get_state_from_status(dev, &new_state, &err_cnt, &status);
+	if (dev_data->state != new_state) {
+		if (dev_data->common.state_change_cb) {
+			dev_data->common.state_change_cb(
+				dev, new_state, err_cnt,
+				dev_data->common.state_change_cb_user_data);
+		}
+
+		if (dev_data->state != CAN_STATE_STOPPED && new_state == CAN_STATE_BUS_OFF) {
+			/* try to recover */
+			XMC_CAN_NODE_ResetInitBit(dev_cfg->can);
```

Comment:
Couldn't this be used when automatic bus recovery is turned off as well?

Reply:
thanks, yes, that's true, I'll work on this tomorrow.

Reply:
@henrikbrixandersen, my original comment was misleading here, sorry.

It's not possible to stop the bus auto-recovery process on the XMC4xxx.
This call to XMC_CAN_NODE_ResetInitBit() is needed to automatically re-enable the node after the bus-recovery completes.

Initially, I thought that we could move this XMC_CAN_NODE_ResetInitBit() call to the driver can_xmc4xxx_recover() function, but it looks tricky to implement because once the internal auto bus recovery completes, it will trigger the can_xmc4xxx_state_change_handler() to indicate that the bus is no longer in a bus-off state. This will then issue a callback to the user about the state change, perhaps, even before they call the recover function.

Reply:
No problem, we can leave the implementation as is for now. I plan on redesigning the auto-recovery/manual recovery functionality soon, effectively turning it into a `can_mode_t` setting. This will allow for drivers to signal whether they support manual recovery at all, which quite a few CAN IPs does not.